### ISSUE TITLE
Idea: Alias shared members with Imports

### ISSUE THREAD

**Issue by gilfusion** [dotnet/vblang#168]
I mentioned this in #103, but it probably deserves its own discussion. I'm not 100% convinced it's a good idea (hence idea instead of a full proposal); there's probably a reason this wasn't implemented before.

# The Idea
Use the existing ```Imports``` alias syntax (```Imports alias = type```) to import shared members (methods, properties, fields, etc.) in addition to types. This would work for both the ```Imports``` statement and imports declared in the project file.

It might mesh well with #162 (Method-scoped Imports).

## Examples
```VB
Imports Print = System.Console.WriteLine
Imports Sqr = System.Math.Sqrt
```

# Possible Use Cases
- Possible way of implementing #103. I can imagine setting up aliases for Print in the in-box project templates, so a new console application could call something like ```Print("Hello, World")``` and it would work.
- Source compatibility with renamed methods. VB6 had Sqr and .NET has Sqrt, for example.

Any others?

# What About Instance Members?
```VB
Imports Print = Foo.WriteLine
Dim Foo As FooClass
Print("Hello, World!")
```
This code would throw a null reference exception on ```Print("Hello, World!")```, where the fact that there is a reference that could be null is hidden. (Maybe there are workarounds to better tie the exception to the Imports statement and the underlying reference without complicating debugging.)

Also, how would you get the instance to invoke the method?
```VB
Imports Print = System.IO.File.AppendText("foo.txt").WriteLine
```
Would every ```Print``` call after this secretly create a new FileStream object? (Maybe the answer to this is just documentation.)

At this point, it's also starting to look a little like macros in disguise. It looks safer to just allow shared members.

Any other thoughts? Does this look like worthwhile at all? Is there a case for any member, or should it be limited to methods?

**Comment by DualBrain** [comment]
I think given your use cases that this is already addressed (IMO better) by creating a reusable library (modules) that add these "keywords".  Don't get me wrong, I like the idea... however, after giving it more thought I don't think it's something that we should pursue given that there is already a solution in place that I think is "better"... meaning we already know what it is and it handles all of the issues that you bring up.  With that said, instead of doing this as part modifying Imports, maybe as part of a Module or Class that there could be a way to short-circuit wrapping a method for this purpose using less lines of code; as long as that approach doesn't seem too confusing... thinking out loud:

Public Sub Print(value As String)
  Console.WriteLine(value)
End Sub

Public Alias System.Console.WriteLine As Print

Wouldn't need Sub/Function as that could be inferred by what is the original method is.  Wouldn't need params as those would be inferred as well.  Public, Private, Friend modifiers seem appropriate.

The problem with this, as I see it, is that this still only works as an "alias"... meaning given my example, Print really isn't PRINT which would be my ultimate goal.  Print would just be an alternative to using WriteLine.  What if I want to have Print do both Write and WriteLine?  That would require (today) to have a method that had an additional param or I'd have to create two different alias which basically puts us in a position of why not just use Write and WriteLine.

So even going down a different rabbit hole, I think it's a great idea to spark conversation... I just don't think that it's something that will ultimately work.  I could be wrong... but I suspect that once this was at the prototype stage it would provide what appears to be great value for very simple scenarios but then break down pretty quickly for anything beyond.

In other words, I do agree that there is a problem here to be solved... I'm just not sure that this is the approach that will do it. ;-)
